1-theory/3-api-contracts.adoc

```diff
@@ -0,0 +1,232 @@
+= API Contracts
+
+The "I" means interface, and so writing a contract for that interface is the act of writing down what that interface is going to look like, to avoid folks having to guess, or go and hunt for some documentation somewhere.
+
+Let's use the fantastic https://pokeapi.co/[PokeAPI] as an example.
+
+[source,javascript]
+----
+{
+  "id": 12,
+  "name": "butterfree",
+  "base_experience": 178,
+  "height": 11,
+  "is_default": true,
+  "order": 16,
+  "weight": 320,
+  "abilities": [{
+    "is_hidden": true,
+    "slot": 3
+  }]
+}
+----
+
+If we look just at the data, we can deduce that sure, ID is probably an auto-incrementing identifier, the name seems to be a lower-cased English string, but after that it starts getting a bit odd.
+
+Height and weight, I guess there is no imperial and metric in the game just units, but what is order all about? The order that pokemon is sat in my current party? That Ash bumped into in the cartoon? In order of cuteness? What?!
+
+An API contract can help answer most of the questions and more for anyone wanting to understand how an API is meant to work.
+
+- Which resources or methods are available and how do you interact with them?
+- Which HTTP endpoints are required for that endpoint, for things like authentication?
+- Which properties are available in the request and response?
+- What are those various properties data types?
+- Which of those properties are required or optional?
+- What other validation rules can be applied to those properties?
+
+[quote, The American Heritage Dictionary of the English Language]
+____
+A contract is an agreement between two or more parties, especially one that is written and enforceable by law.
+____
+
+Thinking about writing an API contract might sound a bit odd if you just think about an API as sharing some data or as a convenient way to trigger commands on another server, but people need to know what that data is, or what arguments they should send to those commands.
+
+The enforceable bit is important too. We can use our contracts in our test suite to confirm that our actual APIs are following the rules of the contract.
+
+NOTE: Some people use the term "API Specifications" to describe all this, but the word "specification" is used to mean a lot of different things. We just used it in the last chapter for something totally different, so let's stick to API Contracts to avoid confusion.
+
+Written down exactly from an early point (and agreed upon) means that there are far fewer surprises throughout the API lifecycle, from planning, through development, and into usage by clients.
```

Comment:
I would move this sentence right after the list below, before `This confidence...`. Seems a bit random after the `NOTE:` bit.